### PR TITLE
fix(cors) properly handle Origin when credentials are enabled

### PR DESCRIPTION
**Pull request by thefosk**
* If `config.credentials=true` and `config.origins=*`, then return the requested `Origin` by the client (if valid).
* If `config.credentials=false` and `config.origins=*`, then return `*` as usual.

### Full changelog

* Better handling of `Access-Control-Allow-Origin` when `config.credentials=true`.

### CODE REVIEW

File: kong/plugins/cors/handler.lua [hunk 1]
```diff
@@ -56,8 +58,24 @@ end
...
+    local allow_credentials = false
+
+    -- If it's catch all, then match the Origin requested by the client
+    -- but only if the Origin URL is valid, and only if credentials=true
+    if conf.credentials then
+      local origin = req_get_headers()["Origin"]
+      if origin then
+        local parsed_url = url.parse(origin)
+        if parsed_url.scheme and parsed_url.host then
+          ngx.header["Access-Control-Allow-Origin"] = fmt("%s://%s",
```
Comment (Tieske) [inline kong/plugins/cors/handler.lua]: In general, don't use uppercase. Openresty returns them all as lowercase. If a header is not found, some metatable magic kicks in converts to lowercase and searches again.

performance wise: always use lowercase

File: kong/plugins/cors/handler.lua [hunk 2]
```diff
@@ -56,8 +58,24 @@ end
...
+        if parsed_url.scheme and parsed_url.host then
+          ngx.header["Access-Control-Allow-Origin"] = fmt("%s://%s", 
+                                                          parsed_url.scheme,
+                                                          parsed_url.host)
+          allow_credentials = true
+        end
+      end
+    end
 
+    ngx.header["Access-Control-Allow-Credentials"] = tostring(allow_credentials)
```
Comment (p0pr0ck5) [inline kong/plugins/cors/handler.lua]: this `tostring` call is unnecessary, the boolean type will be coerced  appropriately.

File: kong/plugins/cors/handler.lua (hunk 1 above)
Comment (p0pr0ck5) [inline kong/plugins/cors/handler.lua]: @Tieske this applies to `ngx.req.get_headers`, not to settings headers via the `ngx.header` API. Consider the following:

```
server {
  listen 12345
  location /bar {
    content_by_lua 'ngx.say("hello world")';
    header_filter_by_lua 'ngx.header["foo"] = "bar"';
  }
}
```

This produces the following result:

```
$ curl -vvv localhost:12345/bar
* Hostname was NOT found in DNS cache
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 12345 (#0)
> GET /bar HTTP/1.1
> User-Agent: curl/7.35.0
> Host: localhost:12345
> Accept: */*
> 
< HTTP/1.1 200 OK
* Server openresty/1.11.2.2 is not blacklisted
< Server: openresty/1.11.2.2
< Date: Fri, 24 Mar 2017 16:54:07 GMT
< Content-Type: application/octet-stream
< Transfer-Encoding: chunked
< Connection: keep-alive
< foo: bar
< 
hello world
```

Assigning the header key as camel case is appropriate here.

Comment (Tieske) [inline kong/plugins/cors/handler.lua]: apologies I meant to put it on line 66.

File: kong/plugins/cors/handler.lua (hunk 2 above)
Comment (Tieske) [inline kong/plugins/cors/handler.lua]: :heart: explicitness

File: kong/plugins/cors/handler.lua [hunk 3]
```diff
@@ -56,8 +58,24 @@ end
 
 local function configure_credentials(ngx, conf)
   if ngx.ctx.cors_allow_all then
-    ngx.header["Access-Control-Allow-Credentials"] = "false"
+    local allow_credentials = false
+
+    -- If it's catch all, then match the Origin requested by the client
+    -- but only if the Origin URL is valid, and only if credentials=true
+    if conf.credentials then
+      local origin = req_get_headers()["Origin"]
```
Comment (thibaultcha) [inline kong/plugins/cors/handler.lua]: As per other occurrences in this plugin, we should be using `ngx.var.http_origin` here.

File: kong/plugins/cors/handler.lua [hunk 4]
```diff
@@ -56,8 +58,24 @@ end
...
+          ngx.header["Access-Control-Allow-Origin"] = fmt("%s://%s", 
+                                                          parsed_url.scheme,
+                                                          parsed_url.host)
+          allow_credentials = true
+        end
+      end
+    end
 
+    ngx.header["Access-Control-Allow-Credentials"] = tostring(allow_credentials)
   elseif conf.credentials then
```
Comment (thibaultcha) [inline kong/plugins/cors/handler.lua]: This file was updated to meet the new code style. As such, we are missing an empty line before this `elseif` statement.

File: kong/plugins/cors/handler.lua (hunk 2 above)
Comment (thibaultcha) [inline kong/plugins/cors/handler.lua]: FYI: this is actually not valid according to the CORS specs. The only accepted value is `true`, case-sensitive.

Comment (thibaultcha) [inline kong/plugins/cors/handler.lua]: ❤️ specifications

File: kong/plugins/cors/handler.lua [hunk 5]
```diff
@@ -56,8 +58,25 @@ end
 
 local function configure_credentials(ngx, conf)
   if ngx.ctx.cors_allow_all then
-    ngx.header["Access-Control-Allow-Credentials"] = "false"
+    local allow_credentials = false
+
+    -- If it's catch all, then match the Origin requested by the client
+    -- but only if the Origin URL is valid, and only if credentials=true
```
Comment (thibaultcha) [inline kong/plugins/cors/handler.lua]: Is there any part of the specification that requires us to only set the ACAO if it is a valid URL?

Comment (thibaultcha) [inline kong/plugins/cors/handler.lua]: i.e., this also "strips" important parts of the URL that would be needed for the client to verify the result of the ACAO header, but now, the client will only receive scheme + host (no authority, no port).

**Comment by thibaultcha** [comment]
Split between and #2451 #2452 with proper git history + updated patch (allowing for port in ACAO header) + simpler logic.

**Pull request closed without merging**